cpp/custrings/text/edit_distance.cu

```diff
@@ -227,3 +231,68 @@ unsigned int NVText::edit_distance( distance_type algo, NVStrings& strs1, NVStri
     return 0;
 }
 
+unsigned int NVText::edit_distance_matrix( distance_type algo, NVStrings& strs, unsigned int* results, bool bdevmem )
+{
+    if( algo != levenshtein || results==0 )
+        throw std::invalid_argument("invalid algorithm");
+    unsigned int count = strs.size();
+    if( count==0 )
+        return 0; // nothing to do
+    auto execpol = rmm::exec_policy(0);
+    // custring_view* d_target = custring_from_host(target);
+
+    // setup results vector
+    unsigned int* d_rtn = results;
+    if( !bdevmem )
+        d_rtn = device_alloc<unsigned int>(count*count,0);
+
+    // get the string pointers
+    rmm::device_vector<custring_view*> strings(count,nullptr);
+    custring_view** d_strings = strings.data().get();
+    strs.create_custring_index(d_strings);
+
+    for (unsigned int k = 0; k < count; k++)
```

Comment:
This can be computed without the CPU loop.
```
Pseudo-code that I think will work:

thrust::for_each_n( 0, count*count,
  [strings, count, results] (idx) {
      row = idx / count;
      col = idx % count;
      if( row < col )
         return;
      dist = 0;  // diagonal
      if( row > col )
         dist = compute_distance(strings[row],strings[col])
      results[idx] = dist; // top half
      transpose_idx = col * count + row;
      results[transpose_idx] = dist; // bottom half
  } );
```

Reply:
There's a good chance it will be faster to just duplicate work for the top and bottom triangles rather than suffer the divergence of the ifs. I have never found it faster to branch in n-body distance kernels.

```
thrust::for_each_n( 0, count*count,
  [strings, count, results] (idx) {
      row = idx / count;
      col = idx % count;
      results[idx] = compute_distance(strings, row, col); // this function can return 0 if row == col
  } );
```